File: sample/sample.cpp
#include <iostream>
#include <string>

#include <xcdat.hpp>

int main() {
    // Dataset of keywords
    std::vector<std::string> keys = {
        "AirPods",  "AirTag",  "Mac",  "MacBook", "MacBook_Air", "MacBook_Pro",
        "Mac_Mini", "Mac_Pro", "iMac", "iPad",    "iPhone",      "iPhone_SE",
    };

    // The input keys must be sorted and unique (although they have already satisfied in this case).
    std::sort(keys.begin(), keys.end());
    keys.erase(std::unique(keys.begin(), keys.end()), keys.end());

    // The trie dictionary type
    using trie_type = xcdat::trie_7_type;
    // using trie_type = xcdat::trie_8_type;
    // using trie_type = xcdat::trie_15_type;
    // using trie_type = xcdat::trie_16_type;

    // The dictionary filename
    const char* tmp_filename = "dic.bin";

    // Build and save the trie dictionary.
    try {
        const trie_type trie(keys);
        xcdat::save(trie, tmp_filename);
    } catch (const xcdat::exception& ex) {
        std::cerr << ex.what() << std::endl;
        return 1;
    }

    // Load the trie dictionary on memory.
    const auto trie = xcdat::load<trie_type>(tmp_filename);

    // Basic statistics
    std::cout << "Number of keys: " << trie.num_keys() << std::endl;
    std::cout << "Number of trie nodes: " << trie.num_nodes() << std::endl;
    std::cout << "Number of DA units: " << trie.num_units() << std::endl;
    std::cout << "Memory usage in bytes: " << xcdat::memory_in_bytes(trie) << std::endl;

    // Lookup the ID for a query key.
    {
        const auto id = trie.lookup("Mac_Pro");
        std::cout << "Lookup(Mac_Pro) = " << id.value_or(UINT64_MAX) << std::endl;
    }
    {
        const auto id = trie.lookup("Google_Pixel");
        std::cout << "Lookup(Google_Pixel) = " << id.value_or(UINT64_MAX) << std::endl;
    }

    // Decode the key for a query ID.
    {
        const auto dec = trie.decode(4);
        std::cout << "Decode(4) = " << dec << std::endl;
    }

    // Common prefix search
    {
        std::cout << "CommonPrefixSearch(MacBook_Air) = {" << std::endl;
        auto itr = trie.make_prefix_iterator("MacBook_Air");
        while (itr.next()) {
            std::cout << "   (" << itr.decoded_view() << ", " << itr.id() << ")," << std::endl;
        }
        std::cout << "}" << std::endl;
    }

    // Predictive search
    {
        std::cout << "PredictiveSearch(Mac) = {" << std::endl;
        auto itr = trie.make_predictive_iterator("Mac");
        while (itr.next()) {
            std::cout << "   (" << itr.decoded_view() << ", " << itr.id() << ")," << std::endl;
        }
        std::cout << "}" << std::endl;
    }

    // Enumerate all the keys (in lex order).
    {
        std::cout << "Enumerate() = {" << std::endl;
        auto itr = trie.make_enumerative_iterator();
        while (itr.next()) {
            std::cout << "   (" << itr.decoded_view() << ", " << itr.id() << ")," << std::endl;
        }
        std::cout << "}" << std::endl;
    }

    std::remove(tmp_filename);

    return 0;
}

File: pybind/sample.py
#!/usr/bin/env python3

import xcdat
import os

# Prepare the dataset of keywords
keyset = xcdat.Keyset()
keyset.append('AirPods')
keyset.append('AirTag')
keyset.append('Mac')
keyset.append('MacBook')
keyset.append('MacBook_Air')
keyset.append('MacBook_Pro')
keyset.append('Mac_Mini')
keyset.append('Mac_Pro')
keyset.append('iMac')
keyset.append('iPad')
keyset.append('iPhone')
keyset.append('iPhone_SE')
keyset.complete()

# You can choose the trie type from the four types.
Trie = xcdat.Trie8
# Trie = xcdat.Trie16
# Trie = xcdat.Trie7
# Trie = xcdat.Trie15

# Build the trie dictionary
trie = Trie(keyset)

# Get the statistics
print(f'Number of keys: {trie.num_keys()}')
print(f'Number of trie nodes: {trie.num_nodes()}')
print(f'Number of DA units: {trie.num_units()}')
print(f'Memory usage in bytes: {trie.memory_in_bytes()}')

# Lookup the ID for a query key.
print(f'Lookup(Mac_Pro) = {trie.lookup("Mac_Pro")}')
print(f'Lookup(Google_Pixel) = {trie.lookup("Google_Pixel")}')

# Decode the key for a query ID.
print(f'Decode(4) = {trie.decode(4)}')

# Common prefix search
print('CommonPrefixSearch(MacBook_Air) = {')
itr = trie.make_prefix_iterator('MacBook_Air')
while itr.next():
    print(f'  ({itr.decoded()}, {itr.id()})')
print('}')

# Predictive search (in lexicographical order).
print('PredictiveSearch(Mac) = {')
itr = trie.make_predictive_iterator('Mac')
while itr.next():
    print(f'  ({itr.decoded()}, {itr.id()})')
print('}')

# Enumerate all the keys (in lexicographical order).
print('Enumerate() = {')
itr = trie.make_enumerative_iterator()
while itr.next():
    print(f'  ({itr.decoded()}, {itr.id()})')
print('}')

# Save the trie to the file
trie.save('dic.bin')

# Load the trie from the file
other = Trie('dic.bin')
assert trie.num_keys() == other.num_keys()
assert trie.num_nodes() == other.num_nodes()
assert trie.num_units() == other.num_units()
assert trie.memory_in_bytes() == other.memory_in_bytes()

os.remove('dic.bin')
